akd/src/directory.rs

```diff
@@ -257,13 +302,60 @@ impl<S: Storage + Sync + Send> Directory<S> {
         }
     }
 
+    /// Poll for changes in the epoch number of the AZKS struct
+    /// stored in the storage layer. If an epoch change is detected,
+    /// the object cache (if present) is flushed immediately so
+    /// that new objects are retrieved from the storage layer against
+    /// the "latest" epoch. There is a "special" flow in the storage layer
+    /// to do a storage-layer retrieval which ignores the cache
+    pub async fn poll_for_azks_changes(
+        &self,
+        period: tokio::time::Duration,
+        change_detected: Option<tokio::sync::mpsc::Sender<()>>,
+    ) -> Result<(), AkdError> {
+        // Retrieve the same AZKS that all the other calls see (i.e. the version that could be cached
+        // at this point). We'll compare this via an uncached call when a change is notified
+        let mut last = Directory::get_azks_from_storage(&self.storage, false).await?;
+
+        loop {
+            // loop forever polling for changes
+            tokio::time::sleep(period).await;
+
+            let latest = Directory::get_azks_from_storage(&self.storage, true).await?;
+            if latest.latest_epoch > last.latest_epoch {
+                {
+                    // acquire a singleton lock prior to flushing the cache to assert that no
+                    // cache accesses are underway (i.e. publish/proof generations/etc)
+                    let _guard = self.cache_lock.write().await;
+                    // flush the cache in its entirety
+                    self.storage.flush_cache().await;
+                    // re-fetch the azks to load it into cache so when we release the cache lock
+                    // others will see the new AZKS loaded up and ready
+                    last = Directory::get_azks_from_storage(&self.storage, false).await?;
```

Comment:
Looks like we have to rely on the storage layer to properly implement cache behavior (in this case remembering to insert into the cache after fetching). I think a rewrite of the storage interface might be necessary in the longer term.

Reply:
So yes and maybe no lol. The storage layer does indeed need to cache, if they want to utilize a cache, however it could also have no cache at all and proxy everything to the storage layer. We've implemented a [`TimedCache`] struct but one could either use their own thing or optionally no cache at all. We could make the storage trait have the default caching logic embedded and just like "native" access calls be the ones to implement by the trait directly...

That could be a future option.